posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,631 @@
+import json
+import os
+from collections.abc import Iterator
+from dataclasses import asdict, dataclass, field
+from datetime import datetime, timedelta, timezone
+from typing import Iterable, NamedTuple
+from uuid import UUID
+
+import psycopg2
+from django.conf import settings
+from temporalio import activity, workflow
+from temporalio.common import RetryPolicy
+
+from posthog.clickhouse.client.execute import sync_execute
+from posthog.temporal.workflows.base import CommandableWorkflow
+
+EPOCH = datetime(1970, 1, 1, 0, 0, tzinfo=timezone.utc)
+
+SELECT_PERSONS_TO_DELETE_QUERY = """
+SELECT
+    team_id,
+    old_person_id,
+    override_person_id,
+    max(created_at) AS latest_created_at,
+    max(version) AS latest_version,
+    min(oldest_event) AS oldest_event_at
+FROM
+    {database}.person_overrides
+WHERE
+    created_at <= %(latest_created_at)s
+GROUP BY
+    team_id, old_person_id, override_person_id
+"""
+
+SELECT_LATEST_CREATED_AT_QUERY = """
+SELECT
+    max(created_at)
+FROM {database}.person_overrides;
+"""
+
+CREATE_JOIN_TABLE_QUERY = """
+CREATE OR REPLACE TABLE {database}.{table_name} (
+    team_id INT NOT NULL,
+    old_person_id UUID NOT NULL,
+    override_person_id UUID NOT NULL
+)
+ENGINE = Join(ANY, LEFT, team_id, old_person_id);
+"""
+
+INSERT_INTO_JOIN_TABLE_QUERY = """
+INSERT INTO {database}.{table_name}
+    SELECT
+        team_id,
+        old_person_id,
+        argMax(override_person_id, created_at) AS override_person_id
+    FROM {database}.person_overrides
+    WHERE created_at <= %(latest_created_at)s
+    GROUP BY team_id, old_person_id;
+"""
+
+SQUASH_EVENTS_QUERY = """
+ALTER TABLE
+    {database}.sharded_events
+UPDATE
+    person_id = joinGet({database}.{table_name}, 'override_person_id', toInt32(team_id), person_id)
+IN PARTITION
+    %(partition_id)s
+WHERE
+    true;
+"""
+
+DROP_JOIN_TABLE_QUERY = """
+DROP TABLE {database}.{table_name};
```

Comment:
This may not work if the join table gets too big
https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#max-table-size-to-drop

Just a failure mode we might need to tune and keep in mind 🧠

Reply:
It should work because it's not a `MergeTree` table - it keeps the data in memory.

From the docs:

_"If the size of a [MergeTree](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree) table exceeds max_table_size_to_drop (in bytes), you can’t delete it using a DROP query."_

However, one failure mode to keep in mind is memory not being freed. Can't recall right now but on some previous CH version one of the in-memory table engines (either this, Memory, or Buffer) didn't free the memory on DROP table. I suspect that's probably been fixed but something for us to keep an eye on.